### ISSUE TITLE
Add support for sidebars in hs.webview

### ISSUE THREAD

**Issue by latenitefilms** [Hammerspoon/hammerspoon#2454]
@asmagill - Would it be possible to "attach" a [sidebar](https://developer.apple.com/design/human-interface-guidelines/macos/windows-and-views/sidebars/) (technically known as a source list) to a `hs.webview` in a similar way that we can attached toolbars? Basically, I'm starting to run out of icon space in my toolbar, so I'm considering replacing the toolbar with a sidebar to allow for more options. Thoughts?

**Related:**
- [Navigating Hierarchical Data Using Outline and Split Views](https://developer.apple.com/documentation/appkit/cocoa_bindings/navigating_hierarchical_data_using_outline_and_split_views?language=objc)
- [PXSourceList](https://github.com/perspx/PXSourceList)

**Comment by asmagill** [comment]
We can easily add the toolbars because NSWindow objects already have built in support for them... there is also some support for title accessories, tab bars, and rulers, but for more complex arrangements in the content section, we're limited at the moment because we're assigning the webview directly to the window's contentView.

*tl;dr - Ok, this got longer than I had anticipated... short version, not easily at present, long version is that I have some ideas about this that might make this, as well as adding other view elements and arrangements, easier to do in the hopefully not too distant future.*

Read the rest if you want some idea of the difficulties and what I'm thinking about trying to implement...
- - -
It's been a while since I looked closely at canvas, webview, and the guitk I've mentioned in the past, but IIRC, the easiest way to do something like this is to put an intervening view between the actual content view (e.g. what contains the actual webview or canvas view) and the window/panel object that houses it. NSSplitView and NSTableView both offer some attractive possibilities, but would require significant additions/changes to canvas and webview to implement at present.

Right now, both `ha.canvas` and `hs.webview` create an `NSPanel` object (NSPanel is a direct subclass of NSWindow, and gives us a few extra options than using NSWindow directly would... for 99% of our purposes, they can be considered interchangeable) and then a content view (NSView for canvas, WKWebView for webview) and assign it as the "contentView" property of the panel. This actually duplicates a fair amount of infrastructure and support code between the two which led to my envisioning a more generic `guitk` set of modules. I think, however, I bit off more than I thought I had when I first started working on it... it has been a daunting challenge to return to it.

I've experimented some with adding subviews to canvas (e.g. they are children of the contentView, not directly of the actual panel) and it's worked relatively well. I haven't advertised it much, as it's somewhat buggy and requires some hacks to do it, but I have successfully put `hs.webview` objects in as children of a canvas, as well as a few of the `guitk` view submodules and a turtle-graphics logo like drawing environment (https://github.com/asmagill/hammerspoon_asm/tree/master/turtle) but I don't really like the shenanigans I have to pull to make it work safely... And because the parent view of these subviews is just another NSView object, we have to handle positioning, scaling, etc. ourselves -- we don't get any of the added benefits that NSTableView or NSSplitView would give us for auto layout, etc. Adding subviews to WKWebView is also theoretically possible, but runs into complications because what we see as a single `WKWebView` object is already an amalgam of internal subviews that are managed dynamically based on the html content... maintaining consistent positioning and layering can sometimes be a challenge; easier to use a view type designed for the purpose (or, in the case of canvas's NSView, designed with very little else going on automatically behind the scenes)

What I have been mulling over lately is to start over with guitk, but start smaller this time...
1. a parent module (currently unnamed) that handles the NSWindow/NSPanel object and *nothing* else. This would primarily consist of the common code between webview and canvas, but put it in one place. Visually it could create a window and background (like needed for a proper webview) or be fully transparent with no window decorations (as used by canvas) but otherwise would provide no visual elements of its own.
2. streamline webview and canvas so that they are just the NSView/WKWebView portion of what they currently represent. These could then be added as the contentView to the parent described in 1. Obviously for backwards compatibility, the default constructor would create the parent object behind the scenes for you, but new constructors would also be added so that item 3 below could be slipped in.
3. NSSplitView, NSTableView, maybe even some of the guitk manager submodules that I envisioned for it could eventually also be added in a support module for 1. These could be assigned to the contentView of 1 and then you would assign the canvas/webview/other-view-objects-yet-to-be-added-to-Hamemrspoon to this intermediate layer.

Separating out the Panel/Window and view objects of canvas and webview would also make using them in other places that can use a view (but don't require or can't even handle the parent object) easier... within status items and submenus, for example (i.e. `hs.menubar`, though probably the replacement version currently attached to my guitk module -- it's got a wrapper for `hs.menubar` and I think is basically ready to go, but I want it in 0.9.80 since we've already got some significant changes in 0.9.79 that need wider exposure first) or the touchbar -- currently the code for this has to separate the canvas view from its panel without releasing it, as most of the canvas methods expect the panel object to exist even when we don't really need or even want it. More complex toolbar items become possible as well... the buttons and input fields of guitk become yet other "views" that can be added to a split or table view, meaning each module/submodule concentrates on it's thing only... we should end up with a variant of my envisioned gui toolkit in phases and yet each piece is smaller and more easily understood and maintained.

Ok, so that last may be a bit of a pipe-dream at the moment, but it's the direction I think this should move, rather than just adding a split view to hs.webview only.

**Comment by latenitefilms** [comment]
Legend, thanks so much for the detailed reply - HUGELY appreciated. That all sounds great.

FWIW - attached is the kind of user interfaces we build for CommandPost. Currently it's just a `hs.webview`, but one day it would be nice to use `hs.guitk` for more control.

<img width="1112" alt="Screen Shot 2020-09-12 at 5 41 16 pm" src="https://user-images.githubusercontent.com/22286696/92990371-607fc900-f51f-11ea-8745-7af80164e066.png">

**Comment by latenitefilms** [comment]
I'm going to close this issue for now, because if the issue does get addressed, it'll most likely be handled in other issues.
